dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -161,56 +163,107 @@ void RtpsUdpDataLink::do_remove_sample(const RepoId& pub_id,
   g.release();
 
   if (writer) {
-    writer->do_remove_sample(criteria);
+    return writer->remove_sample(sample);
   }
+  return REMOVE_NOT_FOUND;
 }
 
-void
-RtpsUdpDataLink::RtpsWriter::do_remove_sample(const TransportQueueElement::MatchCriteria& criteria)
+void RtpsUdpDataLink::remove_all_msgs(const RepoId& pub_id)
 {
-  SnToTqeMap sn_tqe_map;
-  SnToTqeMap to_deliver;
+  ACE_Guard<ACE_Thread_Mutex> g(lock_);
+  RtpsWriter_rch writer;
+  RtpsWriterMap::iterator iter = writers_.find(pub_id);
+  if (iter != writers_.end()) {
+    writer = iter->second;
+  }
 
-  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  g.release();
+
+  if (writer) {
+    writer->remove_all_msgs();
+  }
+}
+
+RemoveResult
+RtpsUdpDataLink::RtpsWriter::remove_sample(const DataSampleElement* sample)
+{
+  bool found = false;
+  SequenceNumber to_release;
+  TransportQueueElement* tqe = 0;
+
+  const char* const payload = sample->get_sample()->cont()->rd_ptr();
+  const TransportQueueElement::MatchOnDataPayload modp(payload);
+
+  ACE_Guard<ACE_Thread_Mutex> g(mutex_);
 
   RtpsUdpDataLink_rch link = link_.lock();
   if (!link) {
-    return;
+    return REMOVE_NOT_FOUND;
   }
 
-  ACE_GUARD(ACE_Thread_Mutex, g2, elems_not_acked_mutex_);
+  RemoveResult result = link->send_strategy()->remove_sample(sample);
+
+  ACE_Guard<ACE_Thread_Mutex> g2(elems_not_acked_mutex_);
 
   if (!elems_not_acked_.empty()) {
-    to_deliver.insert(to_deliver_.begin(), to_deliver_.end());
-    to_deliver_.clear();
-    OPENDDS_SET(SequenceNumber) sns_to_release;
     SnToTqeMap::iterator it = elems_not_acked_.begin();
-    while (it != elems_not_acked_.end()) {
-      if (criteria.matches(*it->second)) {
-        sn_tqe_map.insert(RtpsWriter::SnToTqeMap::value_type(it->first, it->second));
-        sns_to_release.insert(it->first);
-        SnToTqeMap::iterator last = it;
-        ++it;
-        elems_not_acked_.erase(last);
+    while (!found && it != elems_not_acked_.end()) {
```

Comment:
I see that `found` has to be kept for later, but I still wonder if this would be cleaner as a `for` loop with `break` avoiding the unwanted increment instead of needing an `else` block.